CollectionViewHandler2 RelativeSource binding to AncestorType not working

### Description

Using RelativeSource binding AncestorType doesn't seem to work using CollectionViewHandler2

### Steps to Reproduce

1. Add a collecitonview to a page, and bind a property in the itemtemplate to the PageModel using RelativeBinding source of AncestorType:

```xaml
<CollectionView.ItemTemplate>
    <DataTemplate>
        <StackLayout Padding="10">
            <Label Text="{Binding .}" />
            <Button
                Command="{Binding ShowDialogCommand, Source={x:RelativeSource AncestorType={x:Type pm:RelativeSourceBindingPageModel}}}"
                CommandParameter="{Binding .}"
                Text="Show Dialog" />
        </StackLayout>
    </DataTemplate>
</CollectionView.ItemTemplate>
```

2. Enable CollectionViewHandler2:

```cs
#if IOS || MACCATALYST
            builder.ConfigureMauiHandlers(handlers =>
            {
                handlers.AddHandler<Microsoft.Maui.Controls.CollectionView, Microsoft.Maui.Controls.Handlers.Items2.CollectionViewHandler2>();
                handlers.AddHandler<Microsoft.Maui.Controls.CarouselView, Microsoft.Maui.Controls.Handlers.Items2.CarouselViewHandler2>();
            });
#endif
```


3. run the app on iOS and you will see that the binding simply doesn't work 
4. disabling CollectionViewHandler2 resolves the issue

_**See the RelativeSourceBindingPage in the attached reproduction repository**_
CollectionViewHandler2Bugs/Pages/RelativeSourceBindingPage.xaml

### Link to public reproduction project repository

https://github.com/rysbilinski/CollectionViewHandler2Bugs

### Version with bug

9.0.10 SR1

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

8.0.100 SR10

### Affected platforms

iOS

### Affected platform versions

_No response_

### Did you find any workaround?

If you use the normal CollectionViewHandler then the binding works as expected

```cs
 builder
                .UseMauiApp<App>().ConfigureFonts(fonts =>
                {
                    fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
                    fonts.AddFont("OpenSans-Semibold.ttf", "OpenSansSemibold");
                })
                .UseMauiCommunityToolkit();
//#if IOS || MACCATALYST
//            builder.ConfigureMauiHandlers(handlers =>
//            {
//                handlers.AddHandler<Microsoft.Maui.Controls.CollectionView, Microsoft.Maui.Controls.Handlers.Items2.CollectionViewHandler2>();
//                handlers.AddHandler<Microsoft.Maui.Controls.CarouselView, Microsoft.Maui.Controls.Handlers.Items2.CarouselViewHandler2>();
//            });
//#endif
```

### Relevant log output

```shell

```

I can repro this issue at iOS platform on the latest 17.13.0 Preview 1.0(9.0.100 &9.0.0),but I can't verify it on .net8(8.0.100) due to error CS0234.![Image](https://github.com/user-attachments/assets/18722291-e5d3-4bf8-8718-3f062de43f1d)

> I can repro this issue at iOS platform on the latest 17.13.0 Preview 1.0(9.0.100 &9.0.0),but I can't verify it on .net8(8.0.100) due to error CS0234.![Image](https://github.com/user-attachments/assets/18722291-e5d3-4bf8-8718-3f062de43f1d)

won't this be because this handler was inteoduced in .NET 9.0?

--- a/src/Controls/src/Core/Handlers/Items2/iOS/ItemsViewController2.cs
+++ b/src/Controls/src/Core/Handlers/Items2/iOS/ItemsViewController2.cs
@@ -410,7 +410,7 @@ namespace Microsoft.Maui.Controls.Handlers.Items2
 
 				uiView?.Dispose();
 				uiView = null;
-
+				formsElement?.Handler?.DisconnectHandler();
 				formsElement = null;
 			}
 			else

--- a/src/Controls/src/Core/Handlers/Items2/iOS/ItemsViewController2.cs
+++ b/src/Controls/src/Core/Handlers/Items2/iOS/ItemsViewController2.cs
@@ -211,83 +211,6 @@ namespace Microsoft.Maui.Controls.Handlers.Items2
 			CollectionView.ReloadData();
 		}
 
-		// void InvalidateMeasureIfContentSizeChanged()
-		// {
-		// 	var contentSize = CollectionView?.CollectionViewLayout?.CollectionViewContentSize;
-		//
-		// 	if (!contentSize.HasValue)
-		// 	{
-		// 		return;
-		// 	}
-		//
-		// 	bool widthChanged = _previousContentSize.Width != contentSize.Value.Width;
-		// 	bool heightChanged = _previousContentSize.Height != contentSize.Value.Height;
-		//
-		// 	if (_initialized && (widthChanged || heightChanged))
-		// 	{
-		// 		var screenFrame = CollectionView?.Window?.Frame;
-		//
-		// 		if (!screenFrame.HasValue)
-		// 		{
-		// 			return;
-		// 		}
-		//
-		// 		var screenWidth = screenFrame.Value.Width;
-		// 		var screenHeight = screenFrame.Value.Height;
-		// 		bool invalidate = false;
-		//
-		// 		// If both the previous content size and the current content size are larger
-		// 		// than the screen size, then we know that we're already maxed out and the 
-		// 		// CollectionView items are scrollable. There's no reason to force an invalidation
-		// 		// of the CollectionView to expand/contract it.
-		//
-		// 		// If either size is smaller than that, we need to invalidate to ensure that the 
-		// 		// CollectionView is re-measured and set to the correct size.
-		//
-		// 		if (widthChanged && (contentSize.Value.Width < screenWidth || _previousContentSize.Width < screenWidth))
-		// 		{
-		// 			invalidate = true;
-		// 		}
-		//
-		// 		if (heightChanged && (contentSize.Value.Height < screenHeight || _previousContentSize.Height < screenHeight))
-		// 		{
-		// 			invalidate = true;
-		// 		}
-		//
-		// 		if (invalidate)
-		// 		{
-		// 			(ItemsView as IView)?.InvalidateMeasure();
-		// 		}
-		// 	}
-		// 	_previousContentSize = contentSize.Value;
-		// }
-
-		const int HeaderTag = 111;
-
-		// internal Size? GetSize()
-		// {
-		// 	if (_emptyViewDisplayed)
-		// 	{
-		// 		return _emptyUIView.Frame.Size.ToSize();
-		// 	}
-
-		// 	nfloat headerHeight = 0;
-		// 	var headerView = CollectionView.ViewWithTag(HeaderTag);
-
-		// 	if (headerView != null)
-		// 		headerHeight = headerView.Frame.Height;
-
-		// 	var sizeColl = CollectionView.CollectionViewLayout.CollectionViewContentSize;
-		// 	return sizeColl.ToSize();
-		// }
-
-		// void ConstrainItemsToBounds()
-		// {
-		// 	var contentBounds = CollectionView.AdjustedContentInset.InsetRect(CollectionView.Bounds);
-		// 	var constrainedSize = contentBounds.Size;
-		// 	ItemsViewLayout.UpdateConstraints(constrainedSize);
-		// }
-
 		void EnsureLayoutInitialized()
 		{
 			if (_initialized)
@@ -370,7 +293,6 @@ namespace Microsoft.Maui.Controls.Handlers.Items2
 
 				if (!_cellReuseIds.Contains(reuseId))
 				{
-					Console.WriteLine($"REGISTER CELL ID: {reuseId}");
 					CollectionView.RegisterClassForCell(cellType, new NSString(reuseId));
 					_cellReuseIds.Add(reuseId);
 				}
@@ -408,7 +330,6 @@ namespace Microsoft.Maui.Controls.Handlers.Items2
 					//Platform.GetRenderer(formsElement)?.DisposeRendererAndChildren();
 				}
 
-				uiView?.Dispose();
 				uiView = null;
 				formsElement?.Handler?.DisconnectHandler();
 				formsElement = null;

--- a/src/Controls/src/Core/Handlers/Items2/iOS/TemplatedCell2.cs
+++ b/src/Controls/src/Core/Handlers/Items2/iOS/TemplatedCell2.cs
@@ -109,29 +109,28 @@ namespace Microsoft.Maui.Controls.Handlers.Items2
 		public void Bind(DataTemplate template, object bindingContext, ItemsView itemsView)
 		{
 			var virtualView = template.CreateContent(bindingContext, itemsView) as View;
-			BindVirtualView(virtualView, bindingContext, itemsView, true);
+			BindVirtualView(virtualView, bindingContext, itemsView, false);
 		}
 
 		public void Bind(View virtualView, ItemsView itemsView)
 		{
-			BindVirtualView(virtualView, itemsView.BindingContext, itemsView, false);
+			BindVirtualView(virtualView, itemsView.BindingContext, itemsView, true);
 		}
 
-		private void BindVirtualView(View virtualView, object bindingContext, ItemsView itemsView, bool needsContainer)
+		void BindVirtualView(View virtualView, object bindingContext, ItemsView itemsView, bool needsContainer)
 		{
 			if (PlatformHandler is null && virtualView is not null)
 			{
 				var mauiContext = itemsView.FindMauiContext()!;
 				var nativeView = virtualView.ToPlatform(mauiContext);
 
-				if (!needsContainer)
+				if (needsContainer)
 				{
-					PlatformView = nativeView;
+					PlatformView = new UIContainerView2(virtualView, mauiContext);
 				}
 				else
 				{
-					var mauiWrapperView = new UIContainerView2(virtualView, mauiContext);
-					PlatformView = mauiWrapperView;
+					PlatformView = nativeView;
 				}
 
 				PlatformHandler = virtualView.Handler as IPlatformViewHandler;

--- a/src/Controls/src/Core/Handlers/Items2/iOS/TemplatedCell2.cs
+++ b/src/Controls/src/Core/Handlers/Items2/iOS/TemplatedCell2.cs
@@ -64,6 +64,7 @@ namespace Microsoft.Maui.Controls.Handlers.Items2
 			{
 				//view.MeasureInvalidated -= MeasureInvalidated;
 				view.BindingContext = null;
+				(view.Parent as ItemsView)?.RemoveLogicalChild(view);
 			}
 		}
 
@@ -107,49 +108,42 @@ namespace Microsoft.Maui.Controls.Handlers.Items2
 
 		public void Bind(DataTemplate template, object bindingContext, ItemsView itemsView)
 		{
-			if (PlatformHandler is null)
+			var virtualView = template.CreateContent(bindingContext, itemsView) as View;
+			BindVirtualView(virtualView, bindingContext, itemsView, true);
+		}
+
+		public void Bind(View virtualView, ItemsView itemsView)
+		{
+			BindVirtualView(virtualView, itemsView.BindingContext, itemsView, false);
+		}
+
+		private void BindVirtualView(View virtualView, object bindingContext, ItemsView itemsView, bool needsContainer)
+		{
+			if (PlatformHandler is null && virtualView is not null)
 			{
-				var virtualView = template.CreateContent(bindingContext, itemsView) as View;
-
 				var mauiContext = itemsView.FindMauiContext()!;
-				var nativeView = virtualView!.ToPlatform(mauiContext);
+				var nativeView = virtualView.ToPlatform(mauiContext);
 
-				PlatformView = nativeView;
+				if (!needsContainer)
+				{
+					PlatformView = nativeView;
+				}
+				else
+				{
+					var mauiWrapperView = new UIContainerView2(virtualView, mauiContext);
+					PlatformView = mauiWrapperView;
+				}
 
 				PlatformHandler = virtualView.Handler as IPlatformViewHandler;
-
-				InitializeContentConstraints(nativeView);
+				InitializeContentConstraints(PlatformView);
 
 				virtualView.BindingContext = bindingContext;
+				itemsView.AddLogicalChild(virtualView);
 			}
 
 			if (PlatformHandler?.VirtualView is View view)
 			{
 				view.SetValueFromRenderer(BindableObject.BindingContextProperty, bindingContext);
-			}
-		}
-
-		public void Bind(View virtualView, ItemsView itemsView)
-		{
-			if (PlatformHandler is null && virtualView is not null)
-			{
-				var mauiContext = itemsView.FindMauiContext()!;
-				var nativeView = virtualView!.ToPlatform(mauiContext);
-
-				var mauiWrapperView = new UIContainerView2(virtualView, mauiContext);
-
-				PlatformView = mauiWrapperView;
-
-				PlatformHandler = virtualView.Handler as IPlatformViewHandler;
-
-				InitializeContentConstraints(mauiWrapperView);
-
-				virtualView.BindingContext = itemsView.BindingContext;
-			}
-
-			if (PlatformHandler?.VirtualView is View view)
-			{
-				view.SetValueFromRenderer(BindableObject.BindingContextProperty, itemsView.BindingContext);
 			}
 		}
 

--- a/src/Controls/tests/TestCases.HostApp/Issues/Issue26066.xaml
+++ b/src/Controls/tests/TestCases.HostApp/Issues/Issue26066.xaml
@@ -3,7 +3,7 @@
   xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
   xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
   x:Class="Maui.Controls.Sample.Issues.Issue26066"
-  xmlns:cv2="clr-namespace:Maui.Controls.Sample"
+  xmlns:sample="clr-namespace:Maui.Controls.Sample"
   xmlns:local="clr-namespace:Maui.Controls.Sample.Issues"
   x:DataType="local:Issue26066ViewModel">
   <ContentPage.BindingContext>
@@ -27,7 +27,7 @@
       AutomationId="TestLoadButtonCV2"
       Text="Load Items CV2"
       Command="{Binding LoadCommandCV2}"/>
-    <cv2:CollectionView2  Grid.Row="1" BackgroundColor="LightBlue"
+    <sample:CollectionView2  Grid.Row="1" BackgroundColor="LightBlue"
         ItemsSource="{Binding Images2}"
         ItemTemplate="{StaticResource template}"/>
     <Button BackgroundColor="LightGreen "
@@ -35,7 +35,7 @@
       AutomationId="TestLoadButtonCV"
       Text="Load Items CV1"
       Command="{Binding LoadCommand}"/>
-    <CollectionView  Grid.Row="3" BackgroundColor="LightGreen"
+    <sample:CollectionView1  Grid.Row="3" BackgroundColor="LightGreen"
         ItemsSource="{Binding Images}"
         ItemTemplate="{StaticResource template}"/>
   </Grid>

--- a/src/Controls/tests/TestCases.HostApp/Issues/Issue26066.xaml
+++ b/src/Controls/tests/TestCases.HostApp/Issues/Issue26066.xaml
@@ -1,0 +1,42 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<ContentPage
+  xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
+  xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+  x:Class="Maui.Controls.Sample.Issues.Issue26066"
+  xmlns:cv2="clr-namespace:Maui.Controls.Sample"
+  xmlns:local="clr-namespace:Maui.Controls.Sample.Issues"
+  x:DataType="local:Issue26066ViewModel">
+  <ContentPage.BindingContext>
+    <local:Issue26066ViewModel/>
+  </ContentPage.BindingContext>
+  <Grid RowDefinitions="Auto,*,Auto,*">
+    <Grid.Resources>
+      <DataTemplate x:DataType="local:Issue22035Model"
+          x:Key="template">
+        <StackLayout Padding="10" BackgroundColor="Lightgray">
+          <Button
+            AutomationId="{Binding AutomationId}"
+            Command="{Binding ShowDialogCommand, Source={x:RelativeSource AncestorType={x:Type local:Issue26066ViewModel}}, x:DataType=local:Issue26066ViewModel}"
+            CommandParameter="{Binding .}"
+            Text="{Binding Text}"/>
+        </StackLayout>
+      </DataTemplate>
+    </Grid.Resources>
+    <Button BackgroundColor="LightBlue"
+      Grid.Row="0"
+      AutomationId="TestLoadButtonCV2"
+      Text="Load Items CV2"
+      Command="{Binding LoadCommandCV2}"/>
+    <cv2:CollectionView2  Grid.Row="1" BackgroundColor="LightBlue"
+        ItemsSource="{Binding Images2}"
+        ItemTemplate="{StaticResource template}"/>
+    <Button BackgroundColor="LightGreen "
+      Grid.Row="2"
+      AutomationId="TestLoadButtonCV"
+      Text="Load Items CV1"
+      Command="{Binding LoadCommand}"/>
+    <CollectionView  Grid.Row="3" BackgroundColor="LightGreen"
+        ItemsSource="{Binding Images}"
+        ItemTemplate="{StaticResource template}"/>
+  </Grid>
+</ContentPage>

--- a/src/Controls/tests/TestCases.HostApp/Issues/Issue26066.xaml.cs
+++ b/src/Controls/tests/TestCases.HostApp/Issues/Issue26066.xaml.cs
@@ -1,0 +1,42 @@
+﻿using System.Collections.ObjectModel;
+using System.Windows.Input;
+
+namespace Maui.Controls.Sample.Issues
+{
+	[XamlCompilation(XamlCompilationOptions.Compile)]
+	[Issue(IssueTracker.Github, 26066, "CollectionViewHandler2 RelativeSource binding to AncestorType not working", PlatformAffected.Android | PlatformAffected.iOS)]
+	public partial class Issue26066 : ContentPage
+	{
+		public Issue26066()
+		{
+			InitializeComponent();
+		}
+	}
+
+	class Issue26066ViewModel
+	{
+		public Issue26066ViewModel()
+		{
+			LoadCommand.Execute(null);
+			LoadCommandCV2.Execute(null);
+		}
+
+		public ObservableCollection<Issue22035Model> Images { get; set; } = new();
+		public ObservableCollection<Issue22035Model> Images2 { get; set; } = new();
+
+		public ICommand ShowDialogCommand => new Command(async () => await Application.Current.MainPage.DisplayAlert("New Dialog", "Hello from Espinho", "OK"));
+
+		public ICommand LoadCommand => new Command(() => LoadItems(Images, "CV1"));
+
+		public ICommand LoadCommandCV2 => new Command(() => LoadItems(Images2, "CV2"));
+
+		static void LoadItems(ObservableCollection<Issue22035Model> items, string text)
+		{
+			items.Clear();
+			for (int i = 0; i < 3	; i++)
+			{
+				items.Add(new Issue22035Model { Text = $"{text} - Item {i}", ImagePath = i% 2 == 0 ? "photo21314.jpg" : "oasis.jpg" });
+			}
+		}
+	}
+}

--- a/src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue26066.cs
+++ b/src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue26066.cs
@@ -4,13 +4,9 @@ using UITest.Core;
 
 namespace Microsoft.Maui.TestCases.Tests.Issues;
 
-public class Issue26066 : _IssuesUITest
+public class Issue26066(TestDevice testDevice) : _IssuesUITest(testDevice)
 {
-    public Issue26066(TestDevice testDevice) : base(testDevice)
-    {
-    }
-
-    public override string Issue => "CollectionViewHandler2 RelativeSource binding to AncestorType not working";
+	public override string Issue => "CollectionViewHandler2 RelativeSource binding to AncestorType not working";
 
     [Test]
     [Category(UITestCategories.CollectionView)]

--- a/src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue26066.cs
+++ b/src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue26066.cs
@@ -1,0 +1,27 @@
+using NUnit.Framework;
+using UITest.Appium;
+using UITest.Core;
+
+namespace Microsoft.Maui.TestCases.Tests.Issues;
+
+public class Issue26066 : _IssuesUITest
+{
+    public Issue26066(TestDevice testDevice) : base(testDevice)
+    {
+    }
+
+    public override string Issue => "CollectionViewHandler2 RelativeSource binding to AncestorType not working";
+
+    [Test]
+    [Category(UITestCategories.CollectionView)]
+    public void CollectionView2ShouldFindAncestorType()
+    {
+        var text = "CV2";
+        var i = 1;
+        var automationId = $"{text} - Item {i}".Replace(" ", "", StringComparison.Ordinal);
+        App.WaitForElement(automationId);
+        App.Click(automationId);
+        App.WaitForElement("OK");
+        App.Click("OK");
+    }
+}

